### ISSUE TITLE
FR: Adding a Top Level Code Structure

### ISSUE THREAD

**Issue by clarktank** [Dart-Code/Dart-Code#1047]
On Android Studio, on the left side, there is a section called "Structure" where it shows a Top Level of all your classes and methods. Could you bring it to VS code?

**Comment by DanTup** [comment]
I found some screenshots of it, but couldn't figure out exactly what it's showing. Is it not the same as the Outline view?

<img width="390" alt="screen shot 2018-07-02 at 7 41 53 am" src="https://user-images.githubusercontent.com/1078012/42148777-6bf82f48-7dcb-11e8-9f22-c7b0078409f7.png">

**Comment by clarktank** [comment]
@DanTup Yes, that's exactly what I want. How did you enable that feature with Dart or Flutter extension?

**Comment by DanTup** [comment]
It's a VS Code feature - on the Explorer sidebar, right-click and tick `Outline` :) It uses the data we already supplied for the Document Symbol list.

It's currently a preview feature, so it may move - see https://github.com/Microsoft/vscode/issues/49923.

**Comment by clarktank** [comment]
I got it thanks :) It does work on my local files, but for example when I press F12 on **StatefulWidget** to go to its implementation, in this case **framework.dart** file, the outline doesn't show anything!

**Comment by DanTup** [comment]
@clarktank Hmmm, that should work - and seems to here:

<img width="1065" alt="screen shot 2018-07-03 at 4 24 57 pm" src="https://user-images.githubusercontent.com/1078012/42229149-aeffca2c-7edd-11e8-9275-fad897927f9e.png">

Could you try capturing a log? Close `framework.dart`, Run the `Dart: Capture Logs` command, then F12 into the file, give it a few seconds, and then click Stop Logging and attach the log (after reviewing it for any sensitive info/code).

**Comment by clarktank** [comment]
Here is the log file...
[Dart-Code-Log.txt](https://github.com/Dart-Code/Dart-Code/files/2160104/Dart-Code-Log.txt)

**Comment by clarktank** [comment]
Also I am including a screenshot of my VS code project here...
![2018-07-03](https://user-images.githubusercontent.com/31081495/42231178-d6b5d4c8-7eb8-11e8-8dea-141421b9bf14.png)

**Comment by clarktank** [comment]
I don't know if it is useful, but here is a screenshot of my VS code...
![2018-07-03 1](https://user-images.githubusercontent.com/31081495/42231404-74b37d6a-7eb9-11e8-81a1-9ef7a4183ff6.png)

**Comment by DanTup** [comment]
@scheglov Not sure if this is a race (or whether it should work). We opened a file (`framework.dart` inside Flutter) and called `setPriorityFiles` then immediately called `search.getElementDeclarations` (this is VS Code's new Outline feature - I guess it immediately requests the symbols when you open a file). That request happened before the server had started analyzing (from the new file being added as priority) so we ended up with an empty result.

```
[11:47:24 GMT-0400 (Eastern Daylight Time)] [Analyzer] ==> {"id":"121","method":"analysis.setPriorityFiles","params":{"files":["C:\\Users\\clarktank\\Desktop\\Apps\\Flutter\\flutter\\packages\\flutter\\lib\\src\\widgets\\framework.dart","C:\\Users\\clarktank\\Apps\\stephen_1\\lib\\src\\app.dart"]}}
[11:47:24 GMT-0400 (Eastern Daylight Time)] [Analyzer] ==> {"id":"122","method":"analysis.setSubscriptions","params":{"subscriptions":{"CLOSING_LABELS":["C:\\Users\\clarktank\\Desktop\\Apps\\Flutter\\flutter\\packages\\flutter\\lib\\src\\widgets\\framework.dart","C:\\Users\\clarktank\\Apps\\stephen_1\\lib\\src\\app.dart"],"FOLDING":["C:\\Users\\clarktank\\Desktop\\Apps\\Flutter\\flutter\\packages\\flutter\\lib\\src\\widgets\\framework.dart","C:\\Users\\clarktank\\Apps\\stephen_1\\lib\\src\\app.dart"],"OCCURRENCES":["C:\\Users\\clarktank\\Desktop\\Apps\\Flutter\\flutter\\packages\\flutter\\lib\\src\\widgets\\framework.dart","C:\\Users\\clarktank\\Apps\\stephen_1\\lib\\src\\app.dart"],"OUTLINE":["C:\\Users\\clarktank\\Desktop\\Apps\\Flutter\\flutter\\packages\\flutter\\lib\\src\\widgets\\framework.dart","C:\\Users\\clarktank\\Apps\\stephen_1\\lib\\src\\app.dart"]}}}
[11:47:24 GMT-0400 (Eastern Daylight Time)] [Analyzer] ==> {"id":"123","method":"search.getElementDeclarations","params":{"file":"C:\\Users\\clarktank\\Desktop\\Apps\\Flutter\\flutter\\packages\\flutter\\lib\\src\\widgets\\framework.dart"}}
[11:47:24 GMT-0400 (Eastern Daylight Time)] [Analyzer] <== {"id":"121"}
[11:47:24 GMT-0400 (Eastern Daylight Time)] [Analyzer] <== {"id":"122"}
[11:47:24 GMT-0400 (Eastern Daylight Time)] [Analyzer] <== {"event":"server.status","params":{"analysis":{"isAnalyzing":true}}}
[11:47:24 GMT-0400 (Eastern Daylight Time)] [Analyzer] <== {"id":"123","result":{"declarations":[],"files":[]}}
```

I think it would be better if the `setPriorityFiles` request had somehow delayed the `getElementDeclarations` processing until after the re-analysis (though I don't know the complexity or impact of that)?

**Comment by scheglov** [comment]
I don't understand why this might happen.

`search.getElementDeclarations` does not depend on resolution, so it does not need the file to be a priority one. We just need that the file is "known" in one of the drivers / analysis roots, e.g. it is in a root, or in a package in dependencies.

**Comment by DanTup** [comment]
Hmmm, that is strange then. The file must be reachable as it was returned in `getNavigation`. Maybe an instrumentation log will reveal more?

@clarktank Could you try [enabling the Analaysis Server Instrumentation Log](https://dartcode.org/docs/logging/#analyzer-instrumentation), restarting VS Code, reproducing, then providing the log?

Note: That log may include a lot more of your code than the previous log so do review if there might be anything sensitive in there (or you can email it if you don't want to post here - logs @ dartcode .org).

**Comment by clarktank** [comment]
[darm.txt](https://github.com/Dart-Code/Dart-Code/files/2164627/darm.txt)

**Comment by clarktank** [comment]
This time I didn't wait as much as last time, after pressing F12. Please, let men know, if you want me wait more...

**Comment by DanTup** [comment]
@clarktank That's fine, it includes the important part showing the request/response for the data:

```text
1530755229674:Req:{"id"::"30","method"::"search.getElementDeclarations","params"::{"file"::"C::\\Users\\clarktank\\Desktop\\Apps\\Flutter\\flutter\\packages\\flutter\\lib\\src\\widgets\\framework.dart"}}
1530755229676:Res:{"id"::"30","result"::{"declarations"::[],"files"::[]}}
```

@scheglov There's less in that log than I remembered; I thought it showed the files being analyzed. Anything more we can do debug?

**Comment by clarktank** [comment]
I have upgraded to v2.6 and this problem is solved.
